### ISSUE TITLE
Memory leak in torch.compile after aborting backward

### ISSUE THREAD

**Issue by pritamdamania87** [pytorch/pytorch#112090]
### 🐛 Describe the bug

With `torch.compile` enabled, aborting the backward pass seems to still hold onto some memory despite all objects being out of scope and `gc.collect` being called.

The repro script can be found here: https://gist.github.com/pritamdamania87/ee4334010e90a0388f096a39c781b42b. Run as follows:

```
NCCL_ASYNC_ERROR_HANDLING=0 python script.py
```

### Error logs

If we set `compiled = torch.compile(ddp)`:

```
[rank 1] after gc max memory: 44070271488, curr memory: 17668492800
```

If we set `compiled = ddp`:
```
[rank 1] after gc max memory: 44275152896, curr memory: 67108864
```

### Versions

PyTorch 2.0.1

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

**Comment by pritamdamania87** [comment]
On nightly the issue still seems to be there but less pronounced:

```
with torch.compile: [rank 1] after gc max memory: 44070271488, curr memory: 8067730944
without torch.compile: [rank 1] after gc max memory: 44275152896, curr memory: 67108864
```

**Comment by pritamdamania87** [comment]
On 2.1.0:

```
with torch.compile: [rank 1] after gc max memory: 44070271488, curr memory: 8067730944
without torch.compile: [rank 1] after gc max memory: 44275152896, curr memory: 67108864
```

**Comment by pritamdamania87** [comment]
> Is memory freed when you run torch._dynamo.reset()? It is known that the dynamo caches can continue to hold on to references to supposedly freed memory - this could be a related issue.

Will try this out, thanks!

> Also, please report if you get leaks if you try torch.compile with backend="eager" and backend="aot_eager".

I'm just using `torch.compile` with default args. See the repro script: https://gist.github.com/pritamdamania87/ee4334010e90a0388f096a39c781b42b#file-gistfile1-txt-L70

**Comment by williamwen42** [comment]
This is not expected behavior - I'm going to look into fixing it soon.

**Comment by pritamdamania87** [comment]
@williamwen42 Thanks again for looking into this! Was wondering if you had any updates regarding this issue?

**Comment by williamwen42** [comment]
@pritamdamania87 my bad, I got bogged down with other work - I'm going to take a look now.

**Comment by williamwen42** [comment]
Simpler repro:
```python
import gc
import torch

class MyModel(torch.nn.Module):
    def __init__(self):
        super(MyModel, self).__init__()
        self.fc1 = torch.nn.Linear(40000, 30000)
        self.fc2 = torch.nn.Linear(30000, 20000)
        self.fc3 = torch.nn.Linear(20000, 10000)

    def forward(self, out):
        out = self.fc1(out)
        out = self.fc2(out)
        out = self.fc3(out)
        return out

def run(compile):
    mod = MyModel().cuda()
    if compile:
        mod = torch.compile(mod)
    inp = torch.rand(40000, 40000).cuda()
    mod(inp)

def clean_and_report_memory():
    gc.collect()
    print(f"max memory: {torch.cuda.max_memory_allocated()}, curr memory: {torch.cuda.memory_allocated()}")

run(False)
clean_and_report_memory()

run(True)
clean_and_report_memory()
```
gives
```
max memory: 24011460096, curr memory: 8519680
max memory: 24011460096, curr memory: 14409649664
```

**Comment by ezyang** [comment]
I would check if code cache entries are being invalidated when the Module is being deallocated. I bet it isn't. Cc @anijain2305

**Comment by williamwen42** [comment]
It probably has something to do with code cache entries - the lines in `reset()` required to get the memory freed are:
```python
        for weak_code in (
            convert_frame.input_codes.seen + convert_frame.output_codes.seen
        ):
            code = weak_code()
            if code:
                reset_code(code)
        code_context.clear()
```

**Comment by anijain2305** [comment]
> I would check if code cache entries are being invalidated when the Module is being deallocated. I bet it isn't. Cc @anijain2305

In the example provided by @williamwen42 , the `check_fn` gets invalidated when the module goes out of scope. However, invalidation currently is just limited to setting `valid = False` in guards. It doesn't really delete the cache entry. 

@pritamdamania87 Are you unblocked with `torch._dynamo.reset()` workaround for now? We definitely want to solve this but need little bit of work (maybe design) to make it happen. @williamwen42 is prototyping one way to do this.

**Comment by pritamdamania87** [comment]
> @pritamdamania87 Are you unblocked with torch._dynamo.reset() workaround for now? We definitely want to solve this but need little bit of work (maybe design) to make it happen. @williamwen42 is prototyping one way to do this.

We are unblocked with `torch._dynamo.reset()` for now. However, using this option is very painful since it clears out the entire compile cache and requires 30+ mins to compile everything again. Leading to roughly (30 * Ne) mins of total compile times where Ne is the number of times we encounter an error resulting in calls to `torch._dynamo.reset()`. Note that errors are common place in large training jobs.

**Comment by williamwen42** [comment]
The repro is regressing on latest trunk - investigating issue. Will need to modify the added leak test to check for leaks using weakrefs instead of comparing cuda memory allocation numbers.

**Comment by williamwen42** [comment]
Found the cause of the regression here: https://github.com/pytorch/pytorch/pull/120147

**Comment by williamwen42** [comment]
Reverted https://github.com/pytorch/pytorch/pull/120147 and added a stronger test (https://github.com/pytorch/pytorch/pull/120459) in order to prevent a future regression.

**Comment by chauhang** [comment]
@pritamdamania87 Can you please check if you are still seeing this issue with the latest nightlies?
